miden/src/repl/mod.rs

```diff
@@ -0,0 +1,311 @@
+                    if stack_print_flag {
+                        println!("{}", result_string);
+                    }
+                    memory = mem;
+                }
+                Err(e) => {
+                    println!("{}", format!("Error running program: {:?}", e));
+                    program_lines.pop();
+                }
+            }
+        }
+        let readline = rl.readline(">> ");
+        match readline {
+            Ok(line) => {
+                if line == "!program" {
+                    println!("{}", program);
+                    stack_print_flag = false;
+                } else if line == "!help" {
+                    // prints out all the available commands in the Miden Repl tool.
+                    print_instructions();
+                    stack_print_flag = false;
+                } else if line == "!mem" {
+                    for (addr, mem) in &memory {
+                        // each memory register converted from Felt to an integer.
+                        let mem_int = print_mem_address(mem);
+                        println!("{}: {:?}", addr, mem_int)
+                    }
+                    stack_print_flag = false;
+                } else if line.len() > 6 && &line[..5] == "!mem[" {
+                    // if user wants to see the state of a particular address in a memory, the input should be atleast
+                    // of length 5.
+
+                    // extracting the digits from the input text given by the user.
+                    let result = read_mem_address(&line);
+
+                    let mut flag = false;
+
+                    match result {
+                        Ok(addr) => {
+                            for (i, memory_value) in &memory {
+                                if *i == addr {
+                                    // each memory register converted from Felt to an integer.
+                                    let mem_int = print_mem_address(memory_value);
+                                    // displays the state of the memory at index x if present.
+                                    println!("{}: {:?}", addr, mem_int);
+                                    flag = true;
+                                    break;
+                                }
+                            }
+                            if !flag {
+                                println!("Memory at address {} is empty", addr);
+                            }
+                        }
+                        Err(msg) => println!("{}", msg),
+                    }
+
+                    stack_print_flag = false;
+                } else if line == "!undo" {
+                    let last_line = program_lines.pop().unwrap();
+                    println!("Undoing {}", last_line);
+                    stack_print_flag = true;
+                } else if line == "!stack" {
+                    println!("{}", result_string);
+                    stack_print_flag = false;
```

Comment:
i'd use `print_stack()` here as well (once it is modified as described in the comments above). This will also let us get rid of `result_string` variable.

Reply:
I've instead set the stack_print_flag to be true and removed the `println!` statement. The stack will get printed in the next loop when the code enters the `match result` block.